### ISSUE TITLE
Get model metrics after fitting with compute_metrics=False

### ISSUE THREAD

**Issue by AntonLsp** [tensorflow/recommenders#259]
Hello,

I'm training a retrieval model on a lot of data (>800 000 interactions) and a lot of unique items (> 300 000) using precomputed embeddings and contextual data as input.

Because of that large amount of unique items, computing the Top K Accuracy metric is very slow, so i deactivated it.
```
    def __init__(self, user_model, candidate_model, task):
        super().__init__()
        self.candidate_model: tf.keras.Model = candidate_model
        self.user_model: tf.keras.Model = user_model
        self.task: tf.keras.layers.Layer = task
        self.compute_metrics = False

    def compute_loss(self, features, training=False) -> tf.Tensor:
        
        hist, context, label = features
        user_embeddings = self.user_model([hist, context])
        positive_candidates_embeddings = self.candidate_model(label)
        
        # The task computes the loss and the metrics.
        return self.task(user_embeddings, positive_candidates_embeddings, compute_metrics = self.compute_metrics )
```

The problem is that now, I only have the loss value. Even when changing *self.compute_metrics* to true and evaluating the model using *model.evaluate*, I still have my metric to 0.
```
model.compute_metrics = True
results_eval = model.evaluate(val_gen, verbose=0)
print(results_eval)
       
results_eval : [0.0, 463.1528625488281, 0, 463.1528625488281]
```

So here is my question : Is there a way to compute the model's metrics after deactivating it during fitting ?

Thanks for your help !

**Comment by MaiziXiao** [comment]
Do you have a solution for that? I have the same question here. 
Setting compute_metric to True makes the training extremely slow.
I haven't seen a way in the tutorial about how to enable it only for evaluating.  The data in the documentation is too small.

**Comment by AntonLsp** [comment]
Yes, I found a workaround ! Here are the steps :

- In your model, at the end of compute_loss, use `compute_metrics = not training ` when you return your task's loss

- Use None as a metric when defining your task : `task = tfrs.tasks.Retrieval(metrics=None)`

- Train your model as usual

- Define a new metric on your model : `model.task.factorized_metrics = tfrs.metrics.FactorizedTopK(candidates=...), metrics=...)`

- Compile your model `model.compile()`

- `model.evaluate` should now give the proper metrics

**Comment by maciejkula** [comment]
Calling `model.compile()` is the key step here. Keras internally caches the functions used to train, so unless you call `compile` the changes you make to your model will have no effect.

**Comment by MaiziXiao** [comment]
Thanks for the hint. Not sure if I understand correctly but seems to me the metric is computed after each batch during training. Is there a way to calculate the metric only after each epoch?

**Comment by AntonLsp** [comment]
From what i understood, yes, metrics are computed after each batch (same as in keras).
If you want the metric computed after each epoch (for evaluation), just pass `compute_metrics=not training` as argument on your task, at the end of your `compute_loss` function :

```
def compute_loss(self, features: Dict[Text, tf.Tensor], training=False) -> tf.Tensor:
    # We pick out the user features and pass them into the user model.
    user_embeddings = self.user_model(features["user_id"])
    # And pick out the movie features and pass them into the movie model,
    # getting embeddings back.
    positive_movie_embeddings = self.movie_model(features["movie_title"])

    # The task computes the loss and the metrics.
    return self.task(user_embeddings, positive_movie_embeddings, compute_metrics=not training)
```

If you want it only at the end of your training, use the method I described.

**Comment by MaiziXiao** [comment]
Thanks @AntonLsp for the information. I am still quite confused here with the code you pasted. Isn't ` compute_metrics=not training`  equal to ` compute_metrics=True` ? However when I really did the experiment with the code you pasted, the training is indeed much faster. So I think it really does evaluate only at the end of the epoch.
Also, If that's working for you, why are you still trying to evaluate at the end of the training? Is it because the evaluation at the end of each epoch is still too time consuming? (I do realize that the metrics is very slow to compute)

**Comment by AntonLsp** [comment]
`compute_metrics=not training` won't compute metrics during the training step but will during the evaluation step (at the end of each epoch).

I'm only using metrics at the end of the training because calculating metrics after each epoch is still too slow (20 epochs * 3min is still a lot of time lost).  I'm monitoring the training using the loss values (to check under/overfitting) and I only compute metrics at the end, to get only the best ones (I don't care about accuracy at epoch 1 or 2 for example).

**Comment by yunruili** [comment]
Good one. But can I ask @maciejkula where is official document mentioned you can use compute_metrics=not training.
Cannot find in the below
https://www.tensorflow.org/recommenders/api_docs/python/tfrs/tasks/Retrieval

**Comment by patrickorlando** [comment]
Hey @yunruili,
It's an argument that you supply when you call the retrieval layer. 
https://github.com/tensorflow/recommenders/blob/71f85dc0a023f108c09ff4721f526abc62852bb4/tensorflow_recommenders/tasks/retrieval.py#L95-L102

You should set in in the `compute_loss` function of your subclassed `tfrs.Model`.
